python/lsst/ts/m2com/controller_cell.py

```diff
@@ -0,0 +1,475 @@
+        period : `float` or `int`, optional
+            Period to check the telemetry rate in second. (the default is 2)
+        **kwargs : `dict`, optional
+            Additional keyword arguments in "process_telemetry" function
+            call.
+        """
+        self._task_telemetry_loop = asyncio.create_task(
+            self._telemetry_loop(process_telemetry, *args, period=period, **kwargs)
+        )
+
+    async def _telemetry_loop(self, process_telemetry, *args, period=2, **kwargs):
+        """Update and output telemetry information from component.
+
+        Parameters
+        ----------
+        process_telemetry : `func` or `coroutine`
+            Function to process the telemetry. It must has a keyward argument
+            of "message" to receive the message as a dictionary.
+        *args : `args`
+            Arguments needed in "process_telemetry" function call.
+        period : `float` or `int`, optional
+            Period to check the telemetry rate in second. (the default is 2)
+        **kwargs : `dict`, optional
+            Additional keyword arguments in "process_telemetry" function
+            call.
+        """
+
+        self.log.debug("Starting telemetry loop from component.")
+
+        is_coroutine_function = is_coroutine(process_telemetry)
+
+        time_wait_telemetry = 0.0
+        is_telemetry_timed_out = False
+
+        messages_consumed = 0
+        messages_consumed_log_timer = asyncio.create_task(asyncio.sleep(period))
+        while self.run_loops:
+
+            if self.are_clients_connected():
+
+                if (
+                    time_wait_telemetry >= self.TELEMETRY_WAIT_TIMEOUT
+                    and not is_telemetry_timed_out
+                ):
+
+                    self.log.warning(
+                        (
+                            "No telemetry update for more than "
+                            f"{self.TELEMETRY_WAIT_TIMEOUT} seconds. Can you "
+                            f"ping the controller at {self.host} and initiate "
+                            "connection to one of it's ports (at "
+                            f"{self.port_command} and {self.port_telemetry})?"
+                        )
+                    )
+                    time_wait_telemetry = 0.0
+                    is_telemetry_timed_out = True
+
+                # If there is no telemetry, sleep for some time
+                if self.client_telemetry.queue.empty():
+                    await asyncio.sleep(self.timeout)
+                    time_wait_telemetry += self.timeout
+                    continue
+
+                # There is the telemetry to publish
+                time_wait_telemetry = 0.0
+                if is_telemetry_timed_out:
+                    self.log.info("Telemetry is up and running after failure.")
+
+                is_telemetry_timed_out = False
+
+                message = self.client_telemetry.queue.get_nowait()
+
+                # Process the telemetry
```

Comment:
Consider encapsulating these processing steps with a try/except clause and handling any exception.

Reply:
I prefer to expose it and check the possible problems in the unit tests. In addition, the `process_telemetry` can be any function and it would be hard to identify the possible exceptions. Thanks!